web_poet/page_inputs/http.py

```diff
@@ -300,3 +302,14 @@ def _auto_detect_fun(self, body: bytes) -> Optional[str]:
             except UnicodeError:
                 continue
             return resolve_encoding(enc)
+
+
+def request_fingerprint(req: HttpRequest) -> str:
+    """Return the fingerprint of the request."""
+    fp = sha1()
+    fp.update(req.method.encode() + b"\n")
+    fp.update(canonicalize_url(str(req.url)).encode() + b"\n")
+    for name, value in sorted(req.headers.items()):
+        fp.update(f"{name.title()}:{value}\n".encode())
+    fp.update(req.body)
```

Comment:
This seems problematic. Imagine a body that starts with what looks like a header.

Maybe a way to go is to add an extra empty line, like the HTTP protocol does, to separate headers from body.

```suggestion
        fp.update(f"{name.title()}:{value}\n".encode())
    fp.update(b"\n" + req.body)
```